File: backend/src/main/java/com/smartcampus/backend/service/DocumentVerificationService.java
package com.smartcampus.backend.service;

import com.smartcampus.backend.domain.Notification;
import com.smartcampus.backend.domain.OutletApplication;
import com.smartcampus.backend.domain.VerificationReport;
import com.smartcampus.backend.repository.NotificationRepository;
import com.smartcampus.backend.repository.UserRepository;
import com.smartcampus.backend.repository.VerificationReportRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Map;

/**
 * Document Verification Service — 3-layer free approach.
 *
 * Layer 1: Format validation (DocumentFormatValidator — pure regex, instant,
 * free)
 * Layer 2: Public government / bank APIs
 * - GSTN public taxpayer API (free, no key required)
 * - Razorpay IFSC public API (free, no key required)
 * - FoSCoS FSSAI public search (free public government portal)
 * Layer 3: Name cross-matching (our own string logic, free)
 *
 * This service is called asynchronously right after an OutletApplication is
 * saved.
 * The campus admin will see the completed report when they open the review
 * screen.
 *
 * Scoring (out of 100):
 * FSSAI API-verified → 40 pts (format-only fallback → 10 pts)
 * GSTIN API-verified → 30 pts (format-only fallback → 8 pts)
 * PAN format valid → 15 pts
 * IFSC API-verified → 15 pts
 *
 * >= 80 → PASSED (admin can approve confidently)
 * 50–79 → PARTIAL (admin should review manually)
 * < 50 → FAILED (likely invalid/fake documents)
 */
@Service
public class DocumentVerificationService {

    private static final Logger log = LoggerFactory.getLogger(DocumentVerificationService.class);

    private final VerificationReportRepository reportRepo;
    private final NotificationRepository notifRepo;
    private final UserRepository userRepo;
    private final DocumentFormatValidator formatValidator;
    private final RestTemplate restTemplate;

    public DocumentVerificationService(VerificationReportRepository reportRepo,
            NotificationRepository notifRepo,
            UserRepository userRepo,
            DocumentFormatValidator formatValidator,
            RestTemplate restTemplate) {
        this.reportRepo = reportRepo;
        this.notifRepo = notifRepo;
        this.userRepo = userRepo;
        this.formatValidator = formatValidator;
        this.restTemplate = restTemplate;
    }

    // ─────────────────────────────────────────────────────────────────────────
    // ENTRY POINT — called by OutletApplicationController after saving the app
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Kicks off the full verification pipeline asynchronously.
     * Creates a PENDING report immediately, then populates it in the background.
     * The @Async annotation means this returns immediately to the caller —
     * the HTTP response is sent before verification completes.
     */
    @Async
    @Transactional
    public void verifyApplicationAsync(OutletApplication app) {
        log.info("[DocVerify] Starting verification for application id={} outlet='{}'",
                app.getId(), app.getOutletName());

        // Create the report row immediately so the admin can see PENDING status
        VerificationReport report = new VerificationReport(app);
        report = reportRepo.save(report);

        try {
            runVerification(app, report);
        } catch (Exception e) {
            // If anything blows up unexpectedly, mark the report as FAILED
            // so the admin knows something went wrong and can review manually
            log.error("[DocVerify] Unexpected error verifying application id={}: {}",
                    app.getId(), e.getMessage(), e);
            report.setOverallStatus(VerificationReport.STATUS_FAILED);
            report.setFssaiNote("Verification engine error: " + e.getMessage());
            report.setCompletedAt(LocalDateTime.now());
            reportRepo.save(report);
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CORE VERIFICATION LOGIC
    // ─────────────────────────────────────────────────────────────────────────

    private void runVerification(OutletApplication app, VerificationReport report) {

        // ── 1. FSSAI ──────────────────────────────────────────────────────────
        String fssaiNum = app.getFssaiLicenseNumber();
        boolean fssaiFormatOk = fssaiNum != null && formatValidator.validateFssaiFormat(fssaiNum);

        if (!fssaiFormatOk) {
            report.setFssaiVerified(false);
            report.setFssaiNote(fssaiNum == null
                    ? "FSSAI number not provided"
                    : formatValidator.getFssaiFormatError(fssaiNum));
        } else {
            FssaiResult fssai = verifyFssaiViaFoscos(fssaiNum);
            report.setFssaiVerified(fssai.active);
            report.setFssaiRegisteredName(fssai.businessName);
            report.setFssaiExpiryDate(fssai.expiryDate);
            report.setFssaiNote(fssai.note);

            if (fssai.businessName != null) {
                double score = nameMatchScore(app.getOutletName(), fssai.businessName);
                report.setFssaiNameMatchScore(score);
                report.setFssaiNameMismatch(score < 0.5);
                if (score < 0.5) {
                    log.warn("[DocVerify] FSSAI name mismatch: submitted='{}' registered='{}' score={}",
                            app.getOutletName(), fssai.businessName, score);
                }
            }
        }

        // ── 2. GSTIN ─────────────────────────────────────────────────────────
        String gstin = app.getGstin();
        boolean gstFormatOk = gstin != null && formatValidator.validateGstinFormat(gstin);

        if (!gstFormatOk) {
            report.setGstVerified(false);
            report.setGstNote(gstin == null
                    ? "GSTIN not provided"
                    : formatValidator.getGstinFormatError(gstin));
        } else {
            GstResult gst = verifyGstinViaGstn(gstin);
            report.setGstVerified(gst.active);
            report.setGstBusinessName(gst.legalName);
            report.setGstNote(gst.note);

            if (gst.legalName != null) {
                double score = nameMatchScore(app.getOutletName(), gst.legalName);
                report.setGstNameMismatch(score < 0.5);
                if (score < 0.5) {
                    log.warn("[DocVerify] GST name mismatch: submitted='{}' registered='{}' score={}",
                            app.getOutletName(), gst.legalName, score);
                }
            }
        }

        // ── 3. PAN ───────────────────────────────────────────────────────────
        // No free real-time API for PAN — format validation only
        String pan = app.getPanNumber();
        boolean panFormatOk = pan != null && formatValidator.validatePanFormat(pan);
        report.setPanFormatValid(panFormatOk);
        if (pan == null) {
            report.setPanNote("PAN not provided");
        } else {
            report.setPanNote(panFormatOk
                    ? "Format valid (real-time PAN verification requires paid API)"
                    : "Invalid PAN format — expected 5 letters + 4 digits + 1 letter (e.g. ABCDE1234F)");
        }

        // ── 4. Bank / IFSC ───────────────────────────────────────────────────
        String ifsc = app.getBankIfscCode();
        boolean ifscFormatOk = ifsc != null && formatValidator.validateIfscFormat(ifsc);

        if (!ifscFormatOk) {
            report.setBankIfscValid(false);
            report.setBankNote(ifsc == null
                    ? "IFSC code not provided"
                    : "Invalid IFSC format — expected 4 letters + 0 + 6 alphanumeric (e.g. SBIN0001234)");
        } else {
            IfscResult ifscResult = verifyIfscViaRazorpay(ifsc);
            report.setBankIfscValid(ifscResult.valid);
            report.setBankName(ifscResult.bankName);
            report.setBankBranch(ifscResult.branchName);
            report.setBankNote(ifscResult.note);
        }

        // Also validate account number format (no API for this)
        String accNo = app.getBankAccountNumber();
        if (accNo != null && !formatValidator.validateAccountNumberFormat(accNo)) {
            String existing = report.getBankNote();
            report.setBankNote((existing != null ? existing + " | " : "")
                    + "Account number format invalid (expected 9–18 digits)");
        }

        // ── 5. Compute overall score ─────────────────────────────────────────
        int score = 0;

        if (Boolean.TRUE.equals(report.getFssaiVerified()))
            score += 40;
        else if (fssaiFormatOk)
            score += 10; // API unavailable, format OK

        if (Boolean.TRUE.equals(report.getGstVerified()))
            score += 30;
        else if (gstFormatOk)
            score += 8; // API unavailable, format OK

        if (report.isPanFormatValid())
            score += 15;

        if (Boolean.TRUE.equals(report.getBankIfscValid()))
            score += 15;

        report.setOverallScore(score);
        report.setOverallStatus(
                score >= 80 ? VerificationReport.STATUS_PASSED
                        : score >= 50 ? VerificationReport.STATUS_PARTIAL : VerificationReport.STATUS_FAILED);
        report.setCompletedAt(LocalDateTime.now());

        reportRepo.save(report);

        log.info("[DocVerify] Completed for application id={}. Score={} Status={}",
                app.getId(), score, report.getOverallStatus());

        // Notify the campus admin that a new application has been verified
        notifyAdminVerificationComplete(app, report);
    }

    // ─────────────────────────────────────────────────────────────────────────
    // LAYER 2 — Public API calls
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Calls FoSCoS (Food Safety Compliance System) — the official FSSAI portal.
     * Public government endpoint, no API key required.
     * Gracefully falls back to null result if the portal is unreachable.
     */
    @SuppressWarnings("unchecked")
    private FssaiResult verifyFssaiViaFoscos(String licenseNumber) {
        try {
            String url = "https://foscos.fssai.gov.in/api/v1.0/search_lic_reg?lic_no=" + licenseNumber;

            HttpHeaders headers = new HttpHeaders();
            headers.set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36");
            headers.set("Origin", "https://foscos.fssai.gov.in");
            headers.set("Referer", "https://foscos.fssai.gov.in");

            ResponseEntity<Map> response = restTemplate.exchange(
                    url, HttpMethod.GET, new HttpEntity<>(headers), Map.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                Map<String, Object> body = response.getBody();
                String status = (String) body.get("license_status");
                String businessName = (String) body.get("business_name");
                String expiryDate = (String) body.get("expiry_date");

                boolean isActive = "Active".equalsIgnoreCase(status)
                        || "active".equalsIgnoreCase(status);

                return new FssaiResult(
                        isActive,
                        businessName,
                        expiryDate,
                        isActive ? "License is Active" : "License status: " + status);
            }
        } catch (HttpClientErrorException.NotFound e) {
            return new FssaiResult(false, null, null, "FSSAI license number not found in FoSCoS database");
        } catch (Exception e) {
            log.warn("[DocVerify] FoSCoS API unavailable for {}: {}", licenseNumber, e.getMessage());
        }
        // Graceful degradation — API unavailable, format was already validated
        return new FssaiResult(null, null, null,
                "FoSCoS portal unreachable — format validated only. Manual verification recommended.");
    }

    /**
     * Calls the official GSTN (GST Network) taxpayer search endpoint.
     * This is a public government API — no key required, but requires
     * browser-like headers to avoid bot detection.
     */
    @SuppressWarnings("unchecked")
    private GstResult verifyGstinViaGstn(String gstin) {
        try {
            String url = "https://services.gst.gov.in/services/api/search/taxpayerDetails?gstin=" + gstin;

            HttpHeaders headers = new HttpHeaders();
            headers.set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36");
            headers.set("Referer", "https://www.gst.gov.in/");
            headers.set("Accept", "application/json");

            ResponseEntity<Map> response = restTemplate.exchange(
                    url, HttpMethod.GET, new HttpEntity<>(headers), Map.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                Map<String, Object> body = response.getBody();
                String status = (String) body.get("sts"); // "Active" or "Cancelled"
                String legalName = (String) body.get("lgnm"); // legal name
                String tradeName = (String) body.get("tradeNam"); // trade name

                boolean isActive = "Active".equalsIgnoreCase(status);

                // Use trade name if legal name is blank
                String displayName = (legalName != null && !legalName.isBlank())
                        ? legalName
                        : tradeName;

                return new GstResult(
                        isActive,
                        displayName,
                        isActive ? "GST registration is Active" : "GST status: " + status);
            }
        } catch (HttpClientErrorException.NotFound e) {
            return new GstResult(false, null, "GSTIN not found in GSTN database");
        } catch (Exception e) {
            log.warn("[DocVerify] GSTN API unavailable for {}: {}", gstin, e.getMessage());
        }
        return new GstResult(null, null,
                "GSTN portal unreachable — format validated only. Manual verification recommended.");
    }

    /**
     * Calls Razorpay's free public IFSC database API.
     * No authentication required. Returns bank name, branch, address.
     * Returns invalid result (not an exception) when IFSC doesn't exist.
     */
    @SuppressWarnings("unchecked")
    private IfscResult verifyIfscViaRazorpay(String ifsc) {
        try {
            String url = "https://ifsc.razorpay.com/" + ifsc.toUpperCase();
            ResponseEntity<Map> response = restTemplate.getForEntity(url, Map.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                Map<String, Object> body = response.getBody();
                String bank = (String) body.get("BANK");
                String branch = (String) body.get("BRANCH");
                String address = (String) body.get("ADDRESS");

                return new IfscResult(true, bank, branch,
                        "Valid IFSC — " + bank + ", " + branch
                                + (address != null ? " (" + address + ")" : ""));
            }
        } catch (HttpClientErrorException.NotFound e) {
            return new IfscResult(false, null, null,
                    "IFSC code not found in Razorpay's database — verify the code is correct");
        } catch (Exception e) {
            log.warn("[DocVerify] Razorpay IFSC API unavailable for {}: {}", ifsc, e.getMessage());
        }
        return new IfscResult(null, null, null,
                "IFSC API temporarily unavailable — format validated only");
    }

    // ─────────────────────────────────────────────────────────────────────────
    // LAYER 3 — Name cross-matching
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Computes a 0.0–1.0 similarity score between the outlet name submitted
     * by the manager and the registered business name returned by the API.
     *
     * Algorithm: proportion of words in the submitted name that also appear
     * in the registered name (after normalisation).
     *
     * Thresholds:
     * >= 0.7 → Strong match
     * 0.5–0.7 → Likely match
     * < 0.5 → Suspicious mismatch — flag for manual review
     */
    double nameMatchScore(String submittedName, String registeredName) {
        if (submittedName == null || registeredName == null)
            return 0.0;

        String a = normalizeName(submittedName);
        String b = normalizeName(registeredName);

        if (a.equals(b))
            return 1.0;
        if (a.isEmpty() || b.isEmpty())
            return 0.0;

        String[] wordsA = a.split("\\s+");
        if (wordsA.length == 0)
            return 0.0;

        long matchCount = Arrays.stream(wordsA)
                .filter(word -> word.length() > 1) // skip single-char remnants
                .filter(b::contains)
                .count();

        long meaningfulWords = Arrays.stream(wordsA)
                .filter(w -> w.length() > 1)
                .count();

        if (meaningfulWords == 0)
            return 0.0;

        return (double) matchCount / meaningfulWords;
    }

    private String normalizeName(String input) {
        return input.toLowerCase()
                // Remove punctuation
                .replaceAll("[^a-z0-9\\s]", "")
                // Remove common legal suffixes / filler words
                .replaceAll("\\b(pvt|ltd|private|limited|and|the|foods|food|cafe|canteen|kitchen)\\b", "")
                .trim()
                // Collapse multiple spaces
                .replaceAll("\\s+", " ");
    }

    // ─────────────────────────────────────────────────────────────────────────
    // ADMIN NOTIFICATION
    // ─────────────────────────────────────────────────────────────────────────

    private void notifyAdminVerificationComplete(OutletApplication app, VerificationReport report) {
        try {
            // Find the campus admin(s) for this application's campus
            userRepo.findAll().stream()
                    .filter(u -> u.getRole() != null
                            && "ADMIN".equals(u.getRole().getName())
                            && u.getCampus() != null
                            && u.getCampus().getId().equals(app.getCampus().getId()))
                    .forEach(admin -> {
                        String statusEmoji = switch (report.getOverallStatus()) {
                            case VerificationReport.STATUS_PASSED -> "✅";
                            case VerificationReport.STATUS_PARTIAL -> "⚠️";
                            case VerificationReport.STATUS_FAILED -> "❌";
                            default -> "🔄";
                        };

                        String msg = String.format(
                                "New outlet application from %s (%s) has been automatically verified. " +
                                        "Verification result: %s %s (Score: %d/100). " +
                                        "Please review the application and the verification report.",
                                app.getManagerName(),
                                app.getOutletName(),
                                statusEmoji,
                                report.getOverallStatus(),
                                report.getOverallScore());

                        notifRepo.save(new Notification(
                                admin,
                                "New Outlet Application — Verification " + report.getOverallStatus(),
                                msg,
                                Notification.TYPE_OUTLET_APP_APPROVED // reuse closest type
                        ));
                    });
        } catch (Exception e) {
            // Notification failure must never affect the main verification flow
            log.warn("[DocVerify] Failed to notify admin for application id={}: {}",
                    app.getId(), e.getMessage());
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Internal result value objects (private inner records)
    // ─────────────────────────────────────────────────────────────────────────

    /** Null active = API unavailable (degrade gracefully). */
    private record FssaiResult(Boolean active, String businessName, String expiryDate, String note) {
    }

    private record GstResult(Boolean active, String legalName, String note) {
    }

    private record IfscResult(Boolean valid, String bankName, String branchName, String note) {
    }
}

File: backend-fastapi/app/services/document_verification_service.py
"""Document Verification Service — port of ``DocumentVerificationService.java``.

WHAT IS PORTED (faithful to spec §4.11 + INTERVIEW_NOTES Module 2)
------------------------------------------------------------------
The Java original implemented a 3-layer free approach:

    Layer 1: Format validation (pure regex — instant, free)
    Layer 2: Public government / bank APIs
        - FoSCoS FSSAI public search (free, no key)
        - GSTN public taxpayer API (free, no key)
        - Razorpay IFSC public API (free, no key)
    Layer 3: Name cross-matching (own string logic, free)

Per the migration task prompt and spec §4.11 ("format-only, no live government
API call" — explicitly out of scope), THIS PORT IMPLEMENTS LAYER 1 ONLY.
Layers 2 and 3 are NOT wired: every layer-2 lookup returns its existing
"API unavailable → format-only" degraded result (the same graceful-degradation
branch the Java original already had for when those endpoints were unreachable).
This means:

  - The orchestration shape (create PENDING report → run each verification
    step → compute overall score → mark report done) is ported verbatim.
  - The scoring formula (FSSAI=40/10, GST=30/8, PAN=15, IFSC=15) is ported
    verbatim, so a fully-valid set of document formats still scores 33 (10+8+15)
    instead of 100 — exactly as it would in Java when all three APIs were
    unreachable. The admin sees a PARTIAL/FAILED report and reviews manually,
    which is the documented behaviour.
  - The entity shape (``VerificationReport`` columns) is preserved so a real
    verification API can be dropped in later without changing the workflow
    around it (INTERVIEW_NOTES Module 2: "the architecture is specifically
    designed so a real verification API can be dropped in later").

ASYNC vs SYNCHRONOUS (DIVERGENCE FLAGGED)
----------------------------------------
Java: ``verifyApplicationAsync`` was ``@Async @Transactional`` — the controller
returned the saved application immediately, then a background thread created
the PENDING report row and populated it in its own transaction. The campus
admin saw ``status=PENDING`` for a few seconds until the report completed.

Python port: verification runs **synchronously** within the submit request,
in the same transaction as the application save. The report is fully populated
by the time the submit response is returned. This is a documented divergence
flagged in migration-notes/03-applications.md:

  - PRO: deterministic for tests; admin never sees a half-PENDING report;
    simpler than wiring asyncio task dispatch + a separate session.
  - CON: adds the (tiny, regex-only) verification latency to the submit
    response time.

The latency is negligible (pure regex, no I/O) — the divergence is a
simplification, not a behavior change. If real Layer-2 APIs are ever added
later, this is the spot to revisit (use FastAPI ``BackgroundTasks`` or an
asyncio task with its own session).

TRANSACTION MODEL
-----------------
``verify_application`` uses ``db.flush()`` only — the caller owns the commit.
"""
from __future__ import annotations

import logging
from datetime import datetime

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.outlet_application import OutletApplication
from app.models.verification_report import VerificationReport
from app.services import document_format_validator as fv

logger = logging.getLogger(__name__)


# ── Status constants (parity with VerificationReport.java) ────────────────────
STATUS_PENDING = "PENDING"
STATUS_PASSED = "PASSED"
STATUS_PARTIAL = "PARTIAL"
STATUS_FAILED = "FAILED"


async def verify_application(app: OutletApplication, db: AsyncSession) -> VerificationReport:
    """Run Layer-1 document verification for an outlet application.

    Creates a PENDING report row, runs each format check, computes the overall
    score, marks the report completed, flushes — and returns the report. The
    caller owns the commit (same transaction as the application save).

    Any unexpected error is caught and marks the report FAILED with a note,
    exactly like the Java ``runVerification`` catch-all that set the report to
    FAILED and persisted ``"Verification engine error: " + e.getMessage()``.
    """
    # Create the report row immediately so even a partial run leaves a
    # persisted PENDING/FAILED row — parity with Java's
    # ``report = reportRepo.save(new VerificationReport(app))`` before the try.
    report = VerificationReport(
        outlet_application=app,
        fssai_name_mismatch=False,
        gst_name_mismatch=False,
        pan_format_valid=False,
        overall_score=0,
        overall_status=STATUS_PENDING,
        created_at=datetime.now(),
    )
    db.add(report)
    await db.flush()  # populate report.id; keep going inside the same transaction

    try:
        _run_verification(app, report)
    except Exception as e:  # noqa: BLE001 — parity with Java's catch-all
        # Mark the report FAILED so the admin sees something went wrong and
        # can review manually. Mirrors Java's catch-all in verifyApplicationAsync.
        logger.exception(
            "[DocVerify] Unexpected error verifying application id=%s", app.id
        )
        report.overall_status = STATUS_FAILED
        report.fssai_note = f"Verification engine error: {e}"
        report.completed_at = datetime.now()

    await db.flush()
    logger.info(
        "[DocVerify] Completed for application id=%s. Score=%s Status=%s",
        app.id, report.overall_score, report.overall_status,
    )
    return report


def _run_verification(app: OutletApplication, report: VerificationReport) -> None:
    """Populate ``report`` in place with each document's format-only result.

    This is a direct port of ``DocumentVerificationService.runVerification``,
    with the Layer-2 API calls replaced by their existing "API unavailable"
    graceful-degradation branches (spec §4.11: no live government API).

    Scoring (out of 100, ported verbatim):
        FSSAI API-verified → 40 pts (format-only → 10 pts)
        GSTIN API-verified → 30 pts (format-only → 8 pts)
        PAN format valid   → 15 pts
        IFSC API-verified  → 15 pts (format-only → 0 pts — the Java IFSC
                                       branch only credited a real API result)

    The Java IFSC branch credited 15 only when ``bankIfscValid == true`` (the
    real Razorpay lookup) — there was no "format-only → fewer points" fallback
    there. We preserve that asymmetry.
    """
    # ── 1. FSSAI ──────────────────────────────────────────────────────────────
    fssai_num = app.fssai_license_number
    fssai_format_ok = fssai_num is not None and fv.validate_fssai_format(fssai_num)

    if not fssai_format_ok:
        report.fssai_verified = False
        report.fssai_note = (
            "FSSAI number not provided"
            if fssai_num is None
            else fv.get_fssai_format_error(fssai_num)
        )
    else:
        # Layer 2 (FoSCoS) deliberately not wired per spec §4.11. Use the
        # exact "API unavailable" graceful-degradation branch the Java original
        # had for unreachable FoSCoS.
        report.fssai_verified = None
        report.fssai_note = (
            "FoSCoS portal unreachable — format validated only. "
            "Manual verification recommended."
        )

    # ── 2. GSTIN ──────────────────────────────────────────────────────────────
    gstin = app.gstin
    gst_format_ok = gstin is not None and fv.validate_gstin_format(gstin)

    if not gst_format_ok:
        report.gst_verified = False
        report.gst_note = (
            "GSTIN not provided" if gstin is None else fv.get_gstin_format_error(gstin)
        )
    else:
        # Layer 2 (GSTN) deliberately not wired per spec §4.11.
        report.gst_verified = None
        report.gst_note = (
            "GSTN portal unreachable — format validated only. "
            "Manual verification recommended."
        )

    # ── 3. PAN ────────────────────────────────────────────────────────────────
    # No free real-time API for PAN in Java either — format validation only.
    pan = app.pan_number
    pan_format_ok = pan is not None and fv.validate_pan_format(pan)
    report.pan_format_valid = pan_format_ok
    if pan is None:
        report.pan_note = "PAN not provided"
    elif pan_format_ok:
        report.pan_note = "Format valid (real-time PAN verification requires paid API)"
    else:
        report.pan_note = (
            "Invalid PAN format — expected 5 letters + 4 digits + 1 letter "
            "(e.g. ABCDE1234F)"
        )

    # ── 4. Bank / IFSC ────────────────────────────────────────────────────────
    ifsc = app.bank_ifsc_code
    ifsc_format_ok = ifsc is not None and fv.validate_ifsc_format(ifsc)

    if not ifsc_format_ok:
        report.bank_ifsc_valid = False
        report.bank_note = (
            "IFSC code not provided"
            if ifsc is None
            else "Invalid IFSC format — expected 4 letters + 0 + 6 alphanumeric "
                 "(e.g. SBIN0001234)"
        )
    else:
        # Layer 2 (Razorpay IFSC) deliberately not wired per spec §4.11.
        report.bank_ifsc_valid = None
        report.bank_note = "IFSC API temporarily unavailable — format validated only"

    # Account number format (no API for this in Java either)
    acc_no = app.bank_account_number
    if acc_no is not None and not fv.validate_account_number_format(acc_no):
        existing = report.bank_note
        report.bank_note = (
            (existing + " | " if existing else "")
            + "Account number format invalid (expected 9–18 digits)"
        )

    # ── 5. Compute overall score ──────────────────────────────────────────────
    score = 0

    # FSSAI: 40 if API-verified, else 10 if format OK (parity with Java).
    if report.fssai_verified is True:
        score += 40
    elif fssai_format_ok:
        score += 10

    # GST: 30 if API-verified, else 8 if format OK (parity with Java).
    if report.gst_verified is True:
        score += 30
    elif gst_format_ok:
        score += 8

    # PAN: 15 if format valid (no API path exists in Java either).
    if report.pan_format_valid:
        score += 15

    # IFSC: 15 only if API-verified true (Java credited no fallback points here).
    if report.bank_ifsc_valid is True:
        score += 15

    report.overall_score = score
    if score >= 80:
        report.overall_status = STATUS_PASSED
    elif score >= 50:
        report.overall_status = STATUS_PARTIAL
    else:
        report.overall_status = STATUS_FAILED
    report.completed_at = datetime.now()
